fuel-vm/tests/predicate.rs

```diff
@@ -105,3 +105,93 @@ fn get_verifying_predicate() {
         assert!(execute_predicate(predicate, vec![], idx as usize));
     }
 }
+
+fn execute_gas_metered_predicate(predicate: Vec<Opcode>) -> bool {
+    let rng = &mut StdRng::seed_from_u64(2322u64);
+
+    let predicate: Vec<u8> = predicate
+        .into_iter()
+        .flat_map(|op| u32::from(op).to_be_bytes())
+        .collect();
+
+    let amount = 10_000_000;
+
+    let owner = Input::predicate_owner(&predicate);
+    let input = Input::coin_predicate(
+        rng.gen(),
+        owner,
+        amount,
+        AssetId::default(),
+        rng.gen(),
+        0,
+        predicate,
+        vec![],
+    );
+
+    let gas_price = 1_000;
+    let gas_limit = 1_000_000;
+    let script = vec![];
+    let script_data = vec![];
+
+    let mut builder = TransactionBuilder::script(script, script_data);
+    builder.gas_price(gas_price).gas_limit(gas_limit).maturity(0);
+    builder.add_input(input);
+
+    let tx = builder.finalize_checked_basic(0, &ConsensusParameters::default());
+    Interpreter::<PredicateStorage>::check_predicates(tx, Default::default(), Default::default())
+}
+
+fn execute_gas_metered_predicates(predicates: Vec<Vec<Opcode>>) -> bool {
+    let rng = &mut StdRng::seed_from_u64(2322u64);
+
+    let gas_price = 1_000;
+    let gas_limit = 1_000_000;
+    let script = vec![];
+    let script_data = vec![];
+
+    let mut builder = TransactionBuilder::script(script, script_data);
+    builder.gas_price(gas_price).gas_limit(gas_limit).maturity(0);
+
+    for predicate in predicates {
+        let predicate: Vec<u8> = predicate
+            .into_iter()
+            .flat_map(|op| u32::from(op).to_be_bytes())
+            .collect();
+
+        let amount = 10_000_000;
+
+        let owner = Input::predicate_owner(&predicate);
+        let input = Input::coin_predicate(
+            rng.gen(),
+            owner,
+            amount,
+            AssetId::default(),
+            rng.gen(),
+            0,
+            predicate,
+            vec![],
+        );
+
+        builder.add_input(input);
+    }
+
+    let tx = builder.finalize_checked_basic(0, &ConsensusParameters::default());
+    Interpreter::<PredicateStorage>::check_predicates(tx, Default::default(), Default::default())
+}
+
+#[test]
```

Comment:
You could simplify a lot of these tests with `rtest`, since you've already built out helper functions to execute predicates, more tests could easily be added making each predicate an `rstest` testcase like so

```rust
use rstest::rstest;

#[rstest]
#[case(vec![Opcode::RET(REG_ONE)], true)] // This just succeeds
#[case(vec![Opcode::JI(0)], false)] // This runs out of gas
fn predicate_gas_metering(#[case] predicate: Vec<Opcode>, #[case] expected: bool) {
    assert_eq!(execute_gas_metered_predicate(predicate), expected);
}
```

Reply:
I really don't see the benefit here; the current code isn't any less readable in my opinion.

Reply:
I think `rstest` is easier to identify what cases are being tested, and easier to add cases in the future. I know we've been trying to use it more across `fuel-core` too. Main benefit imo though is tests just become more maintainable since adding a test-case (lets say one where a bug is encountered) just becomes 1 line rather than an additional patch test. Technically FWIW each test case can be run in parallel too this way

Reply:
this works fine for single opcode tests but becomes more difficult when constructing larger individual test cases